consensus/istanbul/backend/peer_handler.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package backend
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+type validatorPeerHandler struct {
+	sb *Backend
+}
+
+func (vpl *validatorPeerHandler) AddValidatorPeer(node *enode.Node, address common.Address) {
+	if !vpl.sb.MaintainValConnections() {
+		return
+	}
+
+	// Connect to the remote peer if it's part of the current epoch's valset and
+	// if this node is also part of the current epoch's valset
+
+	block := vpl.sb.currentBlock()
+	valSet := vpl.sb.getValidators(block.Number().Uint64(), block.Hash())
+	if valSet.ContainsByAddress(address) && valSet.ContainsByAddress(vpl.sb.ValidatorAddress()) {
+		vpl.sb.p2pserver.AddPeer(node, p2p.ValidatorPurpose)
```

Comment:
why do we need two operations? 
which probably means writes to 2 channels and they won't be handled atomically on the other side, nor in order. 
not sure if that's a problem.

Reply:
We need to call both since a validator connection must be static (e.g. this node will create an outgoing connection to the other validator if no connection is present) and trusted (e.g. this node will accept the connection from the other validator if it tries to initiate the connection).